### ISSUE TITLE
I need a way to have a Spanner (e.g. Crescendo) start or end with a timestamp instead of a note/chord/rest

### ISSUE THREAD

**Issue by gregchapman-dev** [cuthbertLab/music21#1402]
**Motivation**
Several file formats (MEI, Humdrum) often describe Crescendos/Diminuendos as objects with duration, rather than associating them directly with notes in the same staff.  If there is no note in the same staff with the appropriate start/end time, a parser will need to create an invisible Rest (in a separate Voice) instead, complicating the Score unnecessarily.

**Feature summary**
Add a new object (or perhaps extend the current Dynamic object) to represent a crescendo/diminuendo that is not a Spanner, but instead has duration.

**Proposed implementation**
Simplest bit is defining the new data structure.  Medium bit is coming up with a name.  The trickiest bit will be support for the new object in the MusicXML writer.

**Intent**

[x ] I plan on implementing this myself.
[ ] I am willing to pay to have this feature added.
[ ] I am starting a discussion with the hope that community members will volunteer their time to create this. I understand that individuals work on features of interest to them and that this feature may never be implemented.

**Comment by gregchapman-dev** [comment]
I'm not sure of the v8 schedule; I would love to get this into v8 if at all possible.

**Comment by gregchapman-dev** [comment]
I'll try it out and see what works and doesn't work.  Can a Spanner have no notes?  Just an offset and duration?

**Comment by gregchapman-dev** [comment]
Ah... the first question that needs to be answered, I think, is "how would a crescendo with duration be represented in a MusicXML file"?  And I fear that the duration would advance the time pointer, causing the immediately following note (in the MusicXML) to be offset to the right by the duration of the crescendo.  I think we would actually have to use <backup/> to get the time pointer back to the beginning of the crescendo.  Thoughts?

**Comment by jacobtylerwalls** [comment]
The `<offset>` tag is supposed to take care of all of that, so that the musicxml "time point" can be offset as much as you like. Have you had a look at `setOffsetOptional`? You could also export a 1-measure document straight out of Finale.

**Comment by mscuthbert** [comment]
I believe that the system (using a duration, etc.) would work if at least one note is in the spanner.  I don't know if it works with zero (but can you really crescendo over nothing?)

Definitely hairpins are really under tested in music21.

**Comment by gregchapman-dev** [comment]
setOffsetOptional is almost what I need.  What I need is setOffsetFromWedgeDurationIfThereIsNoEndElement, or something like that.  It's the wedge 'stop' element that needs its offset set, so that it lands at the right spot in the timeline (based on dynWedge.duration).  Poking around still.  I did do some hand-modification of musicxml to test out if direction offset will work, and Finale and Dorico both render correctly; Musescore doesn't.

**Comment by gregchapman-dev** [comment]
Current idea: 
1. Add DynamicWedge.hasOffsetAndDuration (defaults to False)
2. In m21ToXml.py, modify setOffsetOptional to take an offset instead of an object (from which to get the offset).
3. In m21ToXml.py (prePostObjectSpanners), if thisSpanner is a DynamicWedge, and hasOffsetAndDuration, call setOffsetOptional for both 'start' (thisSpanner.offset) and 'stop' (thisSpanner.offset + thisSpanner.quarterLength).

Note that you will still need to put a note from the stream in the DynamicWedge spanner, even though you have set hasOffsetAndDuration, so that it will be processed (even though that note will not be used to derive 'start' or 'stop' offset).

I think that xmlToM21.py will need some work, too, to handle wedge directions with offsets (I think that isn't working right).

**Comment by jacobtylerwalls** [comment]
I was hoping something like this would work, because `self.offsetInMeasure` is already advanced between writing the start and stop wedge tags. WDYT?
```diff
diff --git a/music21/musicxml/m21ToXml.py b/music21/musicxml/m21ToXml.py
index 949d0252f..2486579ef 100644
--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -3283,9 +3283,11 @@ class MeasureExporter(XMLExporterBase):
         '''
         root = self.xmlRoot
         self.objectSpannerBundle = self.spannerBundle.getBySpannedElement(obj)
+        # adjust prePostObjectSpanners to return pairs of spanners and musicxmltags
         preList, postList = self.prePostObjectSpanners(obj)
 
         for sp in preList:  # directions that precede the element
+            self.setOffsetOptional(sp, musicxmltag)
             root.append(sp)
 
         classes = obj.classes
@@ -3333,6 +3335,8 @@ class MeasureExporter(XMLExporterBase):
                 environLocal.printDebug(['did not convert object', obj])
 
         for sp in postList:  # directions that follow the element
+            if sp.quarterLength and opFrac(sp.getLast().offset - sp.getFirst().offset) != sp.quarterLength:
+                self.setOffsetOptional(sp, musicxmltag)
             root.append(sp)
 
     def prePostObjectSpanners(self, target):
```

**Comment by gregchapman-dev** [comment]
Yes, that is very similar to what I have prototyped. 

But I really do need a non-Spanner object (or at least a Spanner I don't have to put anything in).  Here's an example why:

Imagine, if you will, that I wanted to add support for parsing `<hairpin/>` elements in music21's MEI parser.  Here's a simple example from the Verovio test suite: dynam-007.mei...

[dynam-007.mei.zip](https://github.com/cuthbertLab/music21/files/9491587/dynam-007.mei.zip)

Here's the Verovio rendering of this MEI file:

![dynam-007-verovio-rendering](https://user-images.githubusercontent.com/75333244/188500894-59580280-b4b1-439a-89b7-5ca9135785ec.png)

The hairpin element looks like this:

`<hairpin staff="1 2" tstamp="3.000000" tstamp2="0m+4.0000" form="dim" place="between" />`

and it is positioned (in the XML) after the two `<staff/>` elements (which contain all the notes) in the second `<measure/>`.

In order to parse this into a DynamicWedge with spanned notes, I would have to make the parser dive back into both staff elements to (hopefully) find an appropriate note (or two) to put in the DynamicWedge.  Far easier to simply insert (at measure offset 3.0) a Diminuendo object with no spanned notes, but with duration 1.0.

My Humdrum parser will have similar issues, as will music21's MusicXML parser, which will need an update (in this PR) to handle `<wedge/>` 'start's and 'stop's that aren't positioned exactly on a note (either via wedge.offset written by this PR, or via backup/forward, like I managed to get Dorico to write).  The MusicXML case can probably be worked around relatively easily, since there's probably an obvious note to use immediately following the `<wedge/>`, but an MEI and/or Humdrum search for a note is a real pain.

**Comment by jacobtylerwalls** [comment]
Are 0-element spanners filtered out by the musicxml writer?

**Comment by gregchapman-dev** [comment]
That's my assumption, but I will check.  Handling empty DynamicWedges as we iterate through the stream (using offset/duration), and skipping non-empty ones (just handling them as before, when handling the notes that are spanned by them) seems like the right way to go.

**Comment by gregchapman-dev** [comment]
Yes, parseOneElement uses MeasureExporter.classesToMethods and MeasureExporter.wrapAttributeMethodClasses, which don't have entries for things to be skipped (like DynamicWedge/Crescendo/Diminuendo).  I could add it there, and skip it by hand if it has spanned elements.

**Comment by jacobtylerwalls** [comment]
Yeah, let's see a PR. Thanks for "whiteboarding" it with me.

**Comment by gregchapman-dev** [comment]
Working on it.  Thank you!

**Comment by gregchapman-dev** [comment]
PR #1407 is the straw man I'm playing with.  I look forward to your thoughts.

**Comment by gregchapman-dev** [comment]
Hmmm... I think I have failed, and need more ideas. 

I have implemented this, but when I insert a multi-measure diminuendo, things fall apart fairly quickly (the measure containing the diminuendo extends it's duration to fit the whole diminuendo).  music21 really isn't set up to have individual elements cross Measure boundaries unless the element is a Spanner (without a duration, or rather with only an implied duration from the spanned elements).  That seems like a pretty hard wall to break through, yes?  It's a basic assumption of music21.

So my MEI example:

`<hairpin staff="1 2" tstamp="3.000000" tstamp2="0m+4.0000" form="dim" place="between" />`

will parse just fine (tstamp2 is still in the same measure as tstamp), but...

`<hairpin staff="1 2" tstamp="3.000000" tstamp2="2m+4.0000" form="dim" place="between" />`

is going to be a big problem (tstamp2 is more than two measures beyond tstamp).

I think I need to start again.  Any ideas of how to approach this?

**Comment by gregchapman-dev** [comment]
The problem with using invisible rests in DynamicWedges is that they need to be in their own Voice (ugh) because they overlap with the _real_ GeneralNotes, and it really looks a mess, both when rendered and when you are trying to use software to analyze the score.  If there were a music21 object I could use in the spanner, that (unlike invisible rests) would simply be ignored by everyone (except for how it defines the enclosing DynamicWedge's implied offset/duration), then that would be sufficient for me.  If they were constructed to always have duration == 0, I could insert them in the same Voice with the real GeneralNotes, and everything would just work.  Sort of like how a Dynamic('p') can be inserted in a Stream, except without any actual dynamic.

Something like:

`class IgnoreThisObjectEntirely_ItsOnlyHereToBePutInASpanner(Music21Object):`

... although I imagine there's probably a better name.

`class InvisibleNothingWithNoDuration(Music21Object):` or something like that.

**Comment by jacobtylerwalls** [comment]
What happens now if you do that? As in: put a Spanner in m.1 containing a Note in m.1 and a Music21Object in m.2. The musicxml exporter ignores Music21Objects but should still be able to `getLast()` of the spanner -- is it a matter of the `offsetInMeasure` being advanced in the musicxml exporter when writing the spanner? Maybe there is the only change you need -- advance the counter only for the classes that otherwise are allowed to participate in musicxml time tracking, like this condition:

https://github.com/cuthbertLab/music21/blob/00971f4a8e8d44e326e35bd922fd02311bb8f467/music21/musicxml/m21ToXml.py#L3293

**Comment by gregchapman-dev** [comment]
My intention is that these "spanned-only" objects would have zero duration, inserted at the appropriate offset for the spanner endpoint (start or stop).  So I don't think advancing the counter is necessary.  For me, the disadvantage of a bare Music21Object is (1) it's hard to detect such a thing simply via isinstance(obj, Music21Object), since most things would return True, and (2) I'd like to enforce "duration must be zero".

I will test with Music21Object, though, to see how it flies.

**Comment by jacobtylerwalls** [comment]
> For me, the disadvantage of a bare Music21Object is (1) it's hard to detect such a thing simply via isinstance(obj, Music21Object),

Right, you could always use your own subclasses, but if we test with bare `Music21Object` then it shows no further ontology is needed in music21 itself.

**Comment by gregchapman-dev** [comment]
Finally got back to playing with this:

Experiment 1: use bare Music21Object (duration=0).  This doesn't work because m21ToXml.py doesn't position non-GeneralNotes with <forward>.

Experiment 2: use bare GeneralNote(duration=0).  This works in the sense that the <wedge> start and stop are positioned correctly in the output MusicXML, and Finale displays it correctly, but Musescore renders the <wedge> as if stop == start (i.e. the wedge ends up very short indeed).

I will see if I can figure that out, but I'm pleased that I could get reasonable MusicXML out of this.

If you're curious, I am testing using a new version of the MEI parser.  Here is the input MEI file and the output XML file:

[dynam-007.mei.zip](https://github.com/cuthbertLab/music21/files/9977298/dynam-007.mei.zip)
[output.musicxml.zip](https://github.com/cuthbertLab/music21/files/9977303/output.musicxml.zip)

(Note that I don't have a lot of stuff showing up in the correct part yet, just barely implemented enough to try out some experiments.)

**Comment by gregchapman-dev** [comment]
Musescore bug is only "sometimes", apparently, and it hasn't happened again once I got the DynamicWedge start/end into the correct part, so I'm pretty happy with this solution (experiment 2).  I'll write a Musescore bug report at some point if it happens again.

So, @mscuthbert and @jacobtylerwalls are you OK with this approach?  I'm using bare zero-duration GeneralNotes as start and end points of DynamicWedges.  I'm actually putting them up in the Measure (instead of in one of the Voices), which in my opinion makes the MusicXML a bit more readable, since they don't then interfere with the note offsets, etc.  Just a straightforward "let me show you where these weird guys go first, then backup and start the first voice".

**Comment by jacobtylerwalls** [comment]
Sure! And don't forget that you might want your own subclass of GeneralNote.

**Comment by gregchapman-dev** [comment]
I'll do that in my converter, but I'm assuming you don't want that new subclass added to music21.

**Comment by jacobtylerwalls** [comment]
Right, at least in my opinion. But some tests demonstrating that the approach works would be wonderful.

**Comment by gregchapman-dev** [comment]
Will do, once the dust settles.

**Comment by gregchapman-dev** [comment]
It currently would need to be "class SpannerAnchor(GeneralNote)", unless we want to add code to m21ToXml.py to handle it specially.

I would vote for "class SpannerAnchor(Music21Object)" with the code change; I find "class SpannerAnchor(GeneralNote)" weirdly confusing.

But I'm deeply buried in MEI parsing at the moment, so I couldn't take this on.

**Comment by gregchapman-dev** [comment]
Well, never mind, I _do_ volunteer to take this one, because I just ran into a brick wall in this area: parsing TrillExtensions in MEI.

These are Spanners as well, but a TrillExtension with a SpannerAnchor (currently a bare GeneralNote) doesn't get magically handled in m21ToXml.py like DynamicWedges do, because TrillExtension gets parsed as a notation in noteToXML, which is simply not called for bare GeneralNotes.  I tried writing a new spannerAnchorToXml routine (called for bare GeneralNotes for now, but SpannerAnchors someday), but while I can generate mxNotations just fine, I have no mxNote to hang them on.

@mscuthbert and @jacobtylerwalls, I need some suggestions.  First of all, is this even possible in MusicXML?  Can I have a `<wavy-line>` in `<ornaments>` in `<notations>` but _not_ in `<note>`?  Or in some sort of fake `<note>`?  If that's not possible, no mod to music21 will work (for Spanners that are notations: TrillExtension, Glissando, etc).  I would note that I'm quite sure I can find a note to hang the _start_ on, but the end might just be a timestamp that doesn't coincide with a note.

Here's a short MEI file that demonstrates the issue. (Any `<trill>` with an endid or a tstamp2 (a.k.a. end timestamp) gets a wavy line that lasts until the endid note, or the tstamp2 is reached.)  The fourth `<trill>` in this file needs a wavy line that goes to tstamp2=5, which represents the very end of the 4/4 measure.  (tstamp and tstamp2 are encoded in 1-based beats, as described by the currently active time signature's denominator.)

[trills.mei.zip](https://github.com/cuthbertLab/music21/files/10035013/trills.mei.zip)
